src/lib/FreeC/Pass/EtaConversionPass.hs

```diff
@@ -51,45 +97,133 @@
 module FreeC.Pass.EtaConversionPass
   ( etaConversionPass
     -- * Testing interface
+  , etaConvertFuncDecl
   , etaConvertExpr
   )
 where
 
 import           Control.Monad                  ( replicateM )
-import           Data.Maybe                     ( fromMaybe )
+
+import           Data.Maybe                     ( fromMaybe
+                                                , fromJust
+                                                )
 
 import           FreeC.Environment
 import           FreeC.Environment.Fresh
+import           FreeC.Environment.Entry
 import           FreeC.IR.SrcSpan
 import           FreeC.IR.Subterm
 import qualified FreeC.IR.Syntax               as IR
 import           FreeC.Monad.Converter
 import           FreeC.Pass
+-- temporary import; TODO: move this pass before the TypeSignaturePass.
+import           FreeC.Pass.TypeSignaturePass   ( splitFuncType )
+
 
 -- | Applies η-conversions to the right-hand sides of all function declarations
 --   in the given module until all function and constructor applications are
 --   fully applied.
 etaConversionPass :: Pass IR.Module
 etaConversionPass ast = do
-  funcDecls' <- mapM etaConvertFuncDecl (IR.modFuncDecls ast)
+  funcDecls' <- etaConvertFuncDecls (IR.modFuncDecls ast) []
   return ast { IR.modFuncDecls = funcDecls' }
 
 -------------------------------------------------------------------------------
 -- Function declarations                                                     --
 -------------------------------------------------------------------------------
 
--- | Applies 'etaConvertExpr' to the right-hand side of the given function
---   declaration.
+-- | Make sure all occurring functions are fully applied by calling 'etaConvertFuncDecl' on each function declaration. 
+--
+--   If the type of a function declaration changes due to a top-level 
+--   η-conversion, the procedure is performed recursively on all 
+--   previously converted function declarations. 
+--   This ensures that all functions, including mutually-recursive 
+--   functions, are fully applied correctly. 
+etaConvertFuncDecls :: [IR.FuncDecl] -> [IR.FuncDecl] -> Converter [IR.FuncDecl]
+etaConvertFuncDecls []         newFuncDecls = return newFuncDecls
+etaConvertFuncDecls (fd : fds) newFuncDecls = do
+  newFuncDecl <- etaConvertFuncDecl fd
+  if IR.funcDeclReturnType newFuncDecl /= IR.funcDeclReturnType fd
+    then etaConvertFuncDecls (newFuncDecls ++ (newFuncDecl : fds)) []
+    else etaConvertFuncDecls fds (newFuncDecls ++ [newFuncDecl])
+
+-- | Depending on the presence or absence of missing top-level arguments, applies 'modifyTopLevel' or 'etaConvertExpr' to the right-hand side of the given function declaration to ensure all functions 
```

Comment:
This line looks way to long. Make sure no comments exceed the maximum line length of 80 characters.